### ISSUE TITLE
CoreWebView2 crash due to use context menu Open link in new window (WPF)

### ISSUE THREAD

**Issue by Nness** [MicrosoftEdge/WebView2Feedback#2771]
**Description**

When right click on a link and use **Open Link in New Window** may crash the WebView2 process. Once crashed, even if it is recovered by code, every time same action been performed will always crash the `WebView2`. 

![image](https://user-images.githubusercontent.com/1712143/189018509-8cfa6f3d-d567-4afa-9498-edeb332e96e0.png)

When crash, received event from `ProcessFailed`.
![image](https://user-images.githubusercontent.com/1712143/189019769-ab9b7491-166b-4804-84f0-0c0d2b2fb89d.png)

> 2022-09-07 11:50:51.103 +12:00 [ERR] An error has occurred in WebView:: Kind: BrowserProcessExited Reason: Unexpected ExistCode: -1073741819 Description:

**Version**
SDK: 1.0.1020.30 (Also tested on latest version 1.0.1343.22, same outcome)
Runtime: 105.0.1343.27 (Released on 6th of Sep 2022. It also happen on any version before that as well.)
Framework: WPF
OS: Windows 11 Pro, Version 21H2. OS build 22000.856

**Repro Steps**

To reproduce on my own machine by **Open Link in New Window** several times in a row, then I can see the application freeze, and WebView2 start crash.

When open new window with middle mouse click it won't cause crash. I have opened 10 to 20 new tabs by middle mouse button without crash happen. Where use right click context menu, on 4th or 5th attempt it crashed.

**Screenshots**

**Additional context**

#### WebView2 Environment
We use custom environment for all the WebView application creates. 

```C#
 var options = new CoreWebView2EnvironmentOptions {
    AllowSingleSignOnUsingOSPrimaryAccount = true
};

// Under use Roaming folder.
CoreWebView2Environment? webView2Env = await CoreWebView2Environment.CreateAsync(null, paths.LocalData, options);

```

#### Initialize WebView2

We initialize WebView after component initialized.

```XML
        <wv2:WebView2
            Grid.Row="1" Grid.ColumnSpan="2"
            Name="_webView"/>
```

```C#

// WebViewTab.xaml.cs
public WebViewTab(WebViewTabViewModel model) {
    InitializeComponent();
    DataContext = ViewModel = model;
    InitializeWebViewAsync();
}

private async void InitializeWebViewAsync() {
    _webView.NavigationStarting += WebView_NavigationStarting;
    _webView.NavigationCompleted += WebView_NavigationCompleted;
    // On initialization complete, we subscribe to additional event out of CoreWebView2 class. Such as ProcessFailed event and NewWindowRequested event.
    _webView.CoreWebView2InitializationCompleted += WebView_CoreWebView2InitializationCompleted;

   await _webView.EnsureCoreWebView2Async(_configuration.WebViewEnvironment);
}

```

#### Handle new window event

We handle the event by setup `Handled` to be true, and grab the URL and create new UI component ourselves.

```C#

private void HandleNewWindowRequested(object? sender, CoreWebView2NewWindowRequestedEventArgs e) {
    e.Handled = true;
    
    AddNewTab(e.Uri, e.Name);
}

```

---------

After further diagnose, this is related to navigate to new tab right away after open. By delay switch to new tab by 500ms, it resolves the issue. My guess is, the problem is related to context menu. When switch to new view right away, and during context menu close it causes WebView2 crash.

When delay by 500ms before navigate to new tab, it gives window for original context menu to gracefully close.

The middle button click to open new tab will also navigate to new tab, but due to no context menu involved, thus WebView2 crash does not happen.

I will leave this open and feel free to close it. The solution I have found so far works well. But if anyone have better solution feel free to let me know. Thanks.

[AB#43004216](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/43004216)

**Comment by nanaliulei** [comment]
Thanks for reporting! Seems like more data is needed to reproduce this bug. Can you provide more details about what did you do in AddNewTab(e.Uri, e.Name)? That will be very helpful.

**Comment by Nness** [comment]
@nanaliulei here is repository to reproduce the issue. I have tried to make the project as simple as possible.

https://github.com/Nness/WebViewReproduce

If you compile and run the above solution, you should land on Google page. From there, use mouse to right click and open new tab. Then from time to time all of the active WebView will crash.

You can use Visual Studio to pause on [ProcessFailed](https://github.com/Nness/WebViewReproduce/blob/main/WebViewReproduce/WebViewReproduce/TabContent.xaml.cs#L60) to monitor when WebView start to crash.

The crash is always related to right click to open new window. As the image below. `Open Link In New Window`
![image](https://user-images.githubusercontent.com/1712143/210028773-5f3ff95d-402e-4047-a670-9542483b118f.png)

To get around the WebView crash, we have two different ways.

1. Use middle mouse rather than right context menu.
2. On [Handle Event](https://github.com/Nness/WebViewReproduce/blob/main/WebViewReproduce/WebViewReproduce/MainWindow.xaml.cs#L75), we wait 200ms before activate the new tab. 

From my current observation, the context menu is part of WebView, instead of WPF. So, when WebView inactivate too fast, the context menu lost its parents, and it crashes all WebView processes.

That's why, above two alternative way can get away from the crash. Due to 1st one use middle mouse without context menu, where 2nd approach wait 200ms, allow context menu to shutdown before view switches.

**Comment by nanaliulei** [comment]
Thanks for the code you provide, I can reproduce it now. Here is what I found:
1. Open in new window was changed to open in new tab, and the new tab contains a wv2:WebView2 as content. If you use open in new window, the crash won't happen.
2. As you mentioned, add a delay before update the selected tab will stop the crash. if you don't need to go to the new tab immediately, delete that code and let user click the new tab will also work.
3. Open the task manager and you can see the process was running in Efficiency mode. In this mode, webview2 progress won't create until new tab was selected. I tried to run the project with my local build webview2, the process was not in efficiency mode and no error or crash happened. Seems like Efficiency mode delays or affects the close process of context menu too. 

Create in new tab, update selected tab immediately, run in efficiency mode and crash happens. I need to dive deeper with the efficiency mode part. And I will comment on this issue when I found more.

project runs in efficiency mode:
![image](https://user-images.githubusercontent.com/16586920/211472597-885dc729-dde5-482f-b410-8dabdac7c93a.png)

**Comment by Nness** [comment]
I have same conclude as you on first and second one. I didn't notice the efficiency mode makes the difference though. I do know the WebView2 is running under efficiency mode. So far, I am not aware anyway to run it outside of efficiency mode.

The delay works well to avoid crash, but I believe many others will encounter similar issue as well, if they want to achieve something similar to normal browser behaviour. In browser, when click on `target="_blank"` it will open tab, right click can open in either tab or window and focus on it right away. if middle mouse click, then it will open tab without switch over.

Browser's tab behaviour is useful when used for application that need to preserve browsing state across different web app or web site.

It seems Microsoft Edge browser is on efficiency mode as well. Anyway, at moment we are using delay to get around the issue.

**Comment by nanaliulei** [comment]
I understand your needs. But I'm sorry that I have not found a better solution yet. I will continue to work on this. And let you know if I found a new solution.

**Comment by yunate** [comment]
@Nness thanks for your really great feedback!
The issue is really a bug. I am trying to fix it.
The brief reason is the context menu receive lost focus message between MS_DESTROY and MS_NCDESTROY message. But after MS_DESTROY, the destroy of context menu has started, this cause crashed.
So, sleep for 200ms is a workaround, because it will wait until MS_NCDESTROY message received, in MS_NCDESTROY message handle function, it will complete release the context menu.

**Comment by Nness** [comment]
@yunate thanks for update. Good that we have found the bug. 😁

**Comment by champnic** [comment]
This should be fixed in runtimes 112.0.1704.0+. Thanks!
